src/token.rs

```diff
@@ -0,0 +1,221 @@
+        #[derive(Insertable, AsChangeset)]
+        struct NewApiToken<'a> {
+            user_id: i32,
+            name: &'a str,
+        }
+
+        diesel::insert(&NewApiToken {
+            user_id: user_id,
+            name: name,
+        }).into(api_tokens::table)
+            .get_result::<ApiToken>(conn)
+            .map_err(From::from)
+    }
+
+    /// Deletes the provided API token if it belongs to the provided user
+    pub fn delete(conn: &PgConnection, user_id: i32, id: i32) -> CargoResult<()> {
+        diesel::delete(api_tokens::table.find(id).filter(
+            api_tokens::user_id.eq(user_id),
+        )).execute(conn)?;
+        Ok(())
+    }
+
+    pub fn find_for_user(conn: &PgConnection, user_id: i32) -> CargoResult<Vec<ApiToken>> {
+        api_tokens::table
+            .filter(api_tokens::user_id.eq(user_id))
+            .order(api_tokens::created_at.desc())
+            .load::<ApiToken>(conn)
+            .map_err(From::from)
+    }
+
+    pub fn count_for_user(conn: &PgConnection, user_id: i32) -> CargoResult<u64> {
+        api_tokens::table
+            .filter(api_tokens::user_id.eq(user_id))
+            .count()
+            .get_result::<i64>(conn)
+            .map(|count| count as u64)
+            .map_err(From::from)
+    }
+
+    /// Converts this `ApiToken` model into an `EncodableApiToken` for JSON
+    /// serialization.
+    pub fn encodable(self) -> EncodableApiToken {
+        EncodableApiToken {
+            id: self.id,
+            name: self.name,
+            token: None,
+            created_at: ::encode_time(self.created_at),
+            last_used_at: self.last_used_at.map(::encode_time),
+        }
+    }
+
+    /// Converts this `ApiToken` model into an `EncodableApiToken` including
+    /// the actual token value for JSON serialization.
+    pub fn encodable_with_token(self) -> EncodableApiToken {
+        EncodableApiToken {
+            id: self.id,
+            name: self.name,
+            token: Some(self.token),
+            created_at: ::encode_time(self.created_at),
+            last_used_at: self.last_used_at.map(::encode_time),
+        }
+    }
+}
+
+struct BadRequest<T: CargoError>(T);
+
+impl<T: CargoError> CargoError for BadRequest<T> {
+    fn description(&self) -> &str {
+        self.0.description()
+    }
+    fn response(&self) -> Option<Response> {
+        self.0.response().map(|mut response| {
+            response.status = (400, "Bad Request");
+            response
+        })
+    }
+}
+
+impl<T: CargoError> fmt::Display for BadRequest<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Bad Request: {}", self.0)
+    }
+}
+
+fn bad_request<T: CargoError>(error: T) -> Box<CargoError> {
+    Box::new(BadRequest(error))
+}
+
+/// Handles the `GET /me/tokens` route.
+pub fn list(req: &mut Request) -> CargoResult<Response> {
+    let tokens = ApiToken::find_for_user(&*req.db_conn()?, req.user()?.id)?
```

Comment:
What do you think about pulling these out into local variables? I personally find the condensed punctuation hard to read